sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -144,16 +163,172 @@ private[spark] class HiveExternalCatalog(client: HiveClient, hadoopConf: Configu
+          val message =
+            s"Persisting data source table $qualifiedTableName into Hive metastore in" +
+              "Spark SQL specific format, which is NOT compatible with Hive."
+          (None, message)
+
+        // our bucketing is un-compatible with hive(different hash function)
+        case _ if tableDefinition.bucketSpec.nonEmpty =>
+          val message =
+            s"Persisting bucketed data source table $qualifiedTableName into " +
+              "Hive metastore in Spark SQL specific format, which is NOT compatible with Hive. "
+          (None, message)
+
+        case (Some(serde), Some(path)) =>
+          val message =
+            s"Persisting data source table $qualifiedTableName with a single input path " +
+              s"into Hive metastore in Hive compatible format."
+          (Some(newHiveCompatibleMetastoreTable(serde, path)), message)
+
+        case (Some(_), None) =>
+          val message =
+            s"Data source table $qualifiedTableName is not file based. Persisting it into " +
+              s"Hive metastore in Spark SQL specific format, which is NOT compatible with Hive."
+          (None, message)
+
+        case _ =>
+          val provider = tableDefinition.provider.get
+          val message =
+            s"Couldn't find corresponding Hive SerDe for data source provider $provider. " +
+              s"Persisting data source table $qualifiedTableName into Hive metastore in " +
+              s"Spark SQL specific format, which is NOT compatible with Hive."
+          (None, message)
+      }
+
+      (hiveCompatibleTable, logMessage) match {
+        case (Some(table), message) =>
+          // We first try to save the metadata of the table in a Hive compatible way.
+          // If Hive throws an error, we fall back to save its metadata in the Spark SQL
+          // specific way.
+          try {
+            logInfo(message)
+            saveTableIntoHive(table, ignoreIfExists)
+          } catch {
+            case NonFatal(e) =>
+              val warningMessage =
+                s"Could not persist ${tableDefinition.identifier.quotedString} in a Hive " +
+                  "compatible way. Persisting it into Hive metastore in Spark SQL specific format."
+              logWarning(warningMessage, e)
+              saveTableIntoHive(newSparkSQLSpecificMetastoreTable(), ignoreIfExists)
+          }
+
+        case (None, message) =>
+          logWarning(message)
+          saveTableIntoHive(newSparkSQLSpecificMetastoreTable(), ignoreIfExists)
+      }
+    }
+  }
+
+  /**
+   * Converts table schema, partition column names and bucket specification to a (String, String)
+   * map, which will be put into table properties later.
+   */
+  private def tableMetadataToProperties(table: CatalogTable): Map[String, String] = {
+    val properties = new scala.collection.mutable.HashMap[String, String]
+    properties.put(DATASOURCE_PROVIDER, table.provider.get)
```

Comment:
Let's still use getOrElse. Although we will reach here if provider is not hive and the table is not a view, it is still possible that provider is None (e.g. possibly caused by a bug).